src/core/provider.cc

```diff
@@ -801,10 +804,26 @@ DelegatingInferResponseProvider::AllocateOutputBuffer(
       name, content, content_byte_size, content_shape, &output));
 
   if ((output->buffer_ == nullptr) && (content_byte_size > 0)) {
-    char* buffer = new char[content_byte_size];
-    *content = static_cast<void*>(buffer);
-    output->ptr_ = static_cast<void*>(buffer);
-    output->buffer_.reset(buffer);
+    void* buffer = nullptr;
+    TRTSERVER_Error* err = alloc_fn_(
+        allocator_, &buffer, name.c_str(), content_byte_size,
+        TRTSERVER_MEMORY_CPU, 0 /* region_id */, alloc_userp_);
+    if (err != nullptr) {
+      Status status = Status(
+          TrtServerCodeToRequestStatus(TRTSERVER_ErrorCode(err)),
+          TRTSERVER_ErrorMessage(err));
+      TRTSERVER_ErrorDelete(err);
+      return status;
+    }
+    if (buffer == nullptr) {
+      return Status(
+          RequestStatusCode::UNAVAILABLE,
+          "unable to allocate memory for result tensor '" + name + "'");
+    }
+
+    *content = buffer;
+    output->ptr_ = buffer;
+    output->buffer_.reset(reinterpret_cast<char*>(buffer));
```

Comment:
So the allocated memory will be managed by the response provider. For memory that are allocated on GPU, is the smart pointer able to release it properly?

Reply:
No, this change is only an intermediate state for the provider. After I move the grpc and http logic out I plan on simplifying the provider since buffer lifetime management must be handled by the client of TRTSERVER API. There is likely to be some changes in the API to allow that.